nixos-version: output configurationRevision

miallo opened the pull request:
It would be nice to be able to get the `configurationRevision` of the current system.

This allows the user to figure out more information about the generation. 


###### Motivation for this change

It might also be used for #105910 to show additional information for generations in the list of all generations so that the user can have a better understanding of which generation is what.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] Tested, as applicable: => could not find relevant tests
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

miallo committed:
nixos-version: output configurationRevision

`nixos-version --configuration-revision` will show the
configurationRevision.

RaitoBezarius commented:
So sorry it did not get any attention, can you fix the conflicts? I'm interested into getting this merged.

miallo commented:
> So sorry it did not get any attention, can you fix the conflicts? I'm interested into getting this merged.

Absolutely no problem :) I rebased it and switched it from `--configurationRevision` to `--configuration-revision`, since that feels a lot better from my perspective

RaitoBezarius reviewed (approved):
LGTM.

ncfavier commented on nixos/doc/manual/man-nixos-version.xml:
```diff
@@ -118,6 +122,23 @@
+     <para>
+      Show the configuration revision if available. This could be the full SHA1
+      hash of the Git commit of the system flake, if you add
+      <screen>{ system.configurationRevision = if self ? rev then self.rev else "dirty"; }</screen>
```
```suggestion
      <screen>{ system.configurationRevision = self.rev or "dirty"; }</screen>
```

miallo replied:
Much nicer ;)

ncfavier commented on nixos/modules/installer/tools/nixos-version.sh:
```diff
@@ -13,6 +13,13 @@ case "$1" in
     echo "@revision@"
     ;;
+  --configuration-revision)
+    if [[ "@configurationRevision@" =~ "@" ]]; then
```
This is kind of awful, but probably fine for now. I think nixos-version would make a good candidate for `__structuredAttrs` once https://github.com/NixOS/nixpkgs/pull/175649 lands in master.

miallo replied:
I also did not particularly like it - I am happy to take a look into `__structuredAttrs` - should this included in this PR?

ncfavier commented on nixos/modules/installer/tools/nixos-version.sh:
```diff
@@ -13,6 +13,13 @@ case "$1" in
     ;;
+  --configuration-revision)
+    if [[ "@configurationRevision@" =~ "@" ]]; then
+      echo "$0: configuration revision is unknown" >&2
```
Could you fix the other errors to go to stderr as well?

miallo replied:
Yes - I wasn't sure if this was in the scope of this PR, but I think it makes sense :)

miallo committed:
nixos-version: print error for missing revision to stderr

miallo commented:
@RaitoBezarius @ncfavier Do you know who can merge this? Apart from commits by Eelco, the earliest thing apart from treewide shellcheck/shebang changes in this file are from 2016, so I don't really know who to ping...
